design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+  reclaimPolicy: Delete
+```
+
+### Transform Functions
+
+TODO(negz): Explain transform functions.
+
+* Simple, schemafied, functions that can transform claim fields.
+* Similar to Terraform's small set of functions.
+* Purposefully not open ended.
+* Written using plain old YAML.
+
+### Configuration-internal References
+
+Crossplane's [cross resource reference] support must be supplemented in order to
+allow references between dynamically provisioned managed resources to be
+specified in advance. Consider a `ResourceConfigurationClass` that dynamically
+provisions:
+
+* `Subnetwork` A
+* `GKECluster` A
+* `ServiceAccount` A
+* `ServiceAccount` B
+* `GKENodePool` A
+* `GKENodePool` B
+
+The `ResourceConfigurationClass` author would like to configure the resources
+such that:
+
+1. `Subnetwork` A is created in an existing, statically provisioned `Network`.
+1. `GKECluster` A is created in `Subnetwork` A.
+1. `GKENodePool` A uses `ServiceAccount` A.
+1. `GKENodePool` B uses `ServiceAccount` B.
+1. Both `GKENodePool` resources join `GKECluster` A.
+
+The author cannot use a contemporary cross resource reference for requirements
+two through five. Managed resources are referenced by name, and the names of
+dynamically provisioned resources are not known until they have been
+provisioned. Thus the author must be able to configure rules such as:
+
+* Managed resources that are dynamically provisioned using resource class A
+  should reference a managed resource that was dynamically provisioned using
+  resource class B.
+* Managed resources that are dynamically provisioned using resource class A
+  should reference a managed resource that is part of the same
+  `ResourceConfiguration`.
+
+This document proposes the introduction of a reference _selector_:
+
+```yaml
+apiVersion: compute.gcp.crossplane.io/v1alpha3
+kind: Subnetwork
+metadata:
+  name: coolsubnet
+spec:
+  forProvider:
+    network: /projects/foo/global/networks/bar
+    networkRef:
+      name: foobar
+    networkSelector:
+      matchConfiguration: true
+      matchLabels:
+        example: label
+```
+
+The above example shows a GCP `Subnetwork` managed resource that uses a
+reference selector. `networkSelector` sets the `networkRef` which in turn sets
+the value of `network`. The reference selector contains two optional fields:
+
+* `matchConfiguration`: Match only managed resources that are part of the same
+  `ResourceConfiguration` (i.e. have the same `configurationRef`) as the
+  selecting resource.
+* `matchLabels`: Match only managed resources with the supplied labels.
+
+If a reference field is set, its corresponding selector field is ignored. If the
+selector field is unset, it is ignored. If the specified selector matches
+multiple managed resources one is chosen at random. Note however that setting
+both `matchConfiguration` and `matchLabels` can guarantee that at most one
+provisioned managed resource will match the selector.
```

Comment:
This could get somewhat complicated with large configurations. I think we would likely need some sort of tooling around authoring the configuration classes such that a user can quickly validate how a configuration will shake out.

Reply:
How large do you think configurations would be likely to get? We've flirted with the idea of what I'm calling a configuration being somewhat similar to a Terraform module, but I have a hard time imagining folks packaging more than maybe 4-5 resources together into a configuration.

Do you think there's a different approach that could be simpler?

Reply:
I think I agree, and even if someone does try to put much more into a large configuration, that is arguably just a poor design choice by a user. Thinking of terraform modules, I am realizing that this model doesn't really allow for layering of `ResourceConfigurationClass` (or maybe I am mistaken?). Do you have any thoughts about allowing a `ResourceConfigurationClass` to reference other `ResourceConfigurationClasses` such that you could have something like a GKE cluster with node pools `ResourceConfigurationClass` that could be referenced by another `ResourceConfigurationClass` that was even more abstract, like `ApplicationXDeployment` that also wants a GKE cluster with node pools but will fill in some of the values for you?

Reply:
I don't think you could use a `ResourceConfigurationClass` inside another `ResourceConfigurationClass` under this design, because `ResourceConfigurationClass` embeds managed resources, not resource classes. So you'd have to embed a `ResourceConfiguration`, which under this proposal just aggregates a bunch of managed resources rather than defining their configuration.

I feel like whether we _want_ to support nesting resource configurations has parallels to my wondering in https://github.com/crossplaneio/crossplane/pull/1163/files#r376261078 around whether we're thinking of this `ResourceConfiguration` as more of a "composite managed resource" (singular) or a something more like a Terraform module. Put otherwise, is the use case more "this resource configuration represents a database, but a useful database happens to require creating 3 different external resources", or more "this resource configuration represents all of my production infrastructure", or "this resource configuration represents all the infrastructure needs of my application".

My feeling is that we should keep it simple(ish) here and solve for the former, which seems like the concrete problem we have today.

Reply:
I'm preeeeetty sure this discussion is defunct under the latest iteration of this design, so I'm going to mark it resolved.